engine/src/main/java/org/camunda/bpm/engine/impl/batch/removaltime/ProcessSetRemovalTimeJobHandler.java

```diff
@@ -84,24 +122,30 @@ protected Date getOrCalculateRemovalTime(SetRemovalTimeBatchConfiguration batchC
     }
   }
 
-  protected void addRemovalTimeToHierarchy(String rootProcessInstanceId, Date removalTime, CommandContext commandContext) {
-    commandContext.getHistoricProcessInstanceManager()
-      .addRemovalTimeToProcessInstancesByRootProcessInstanceId(rootProcessInstanceId, removalTime);
-
-    if (isDmnEnabled(commandContext)) {
-      commandContext.getHistoricDecisionInstanceManager()
-        .addRemovalTimeToDecisionsByRootProcessInstanceId(rootProcessInstanceId, removalTime);
+  protected UpdateResult addRemovalTimeToHierarchy(String rootProcessInstanceId, Date removalTime, UpdateContext updateContext, CommandContext commandContext) {
+    UpdateResult updateResult = commandContext.getHistoricProcessInstanceManager()
+        .addRemovalTimeToProcessInstancesByRootProcessInstanceId(rootProcessInstanceId, removalTime, updateContext);
+    if (updateResult.remainingBatchSize > 0 && isDmnEnabled(commandContext)) {
+      // process tables are done, update DMN tables for the same instance
+      return commandContext.getHistoricDecisionInstanceManager()
+          .addRemovalTimeToDecisionsByRootProcessInstanceId(rootProcessInstanceId, removalTime, updateContext);
```

Comment:
❌ The process table index needs to be reflected in the result by the DMN table update, otherwise a potential next run for the same instance wouldn't know that the process tables are done already.